### PR TITLE
AlwaysOn Profiller - shared generic classes and internal classes

### PR DESCRIPTION
**Pull request by Kielek**
## Why

Extend the AlwaysOn Profiler.

## What

Support for internal classed and shared generic functions.
Cache for class names is removed. Full function names extended about the class name size.

## Tests

CI

## TODO
You can expect more PRs for handling generic parameters.

### CODE REVIEW

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp [hunk 1]
```diff
@@ -366,43 +304,43 @@ class SamplingHelper
...
         }
 
-        // If the ClassID returned from GetFunctionInfo is 0, then the function
-        // is a shared generic function.
-        if (classId != 0)
-        {
-            LookupClassName(classId, result);
-        }
-        else
+        if (function_info.type.parent_type != nullptr)
```
Comment (pjanotti) [inline tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:320]: For which cases `parent_type` comes into the picture?

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp [hunk 2]
```diff
@@ -246,13 +242,12 @@ class SamplingHelper
     // These are permanent parts of the helper object
     ICorProfilerInfo10* info10 = nullptr;
     NameCache functionNameCache;
-    NameCache classNameCache;
     // These cycle every sample and/or are owned externally
     ThreadSamplesBuffer* curWriter = nullptr;
     std::vector<unsigned char>* curBuffer = nullptr;
     SamplingStatistics stats;
 
-    SamplingHelper() : functionNameCache(max_function_name_cache_size), classNameCache(max_class_name_cache_size)
```
Comment (pjanotti) [inline tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:255]: 👍

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp (hunk 1 above)
Comment (pjanotti) [reply tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:320]: Is it for the case like `My.Custom.Test.Namespace.ClassA.InternalClassB.DoubleInternalClassB`? Perhaps it is worth a comment to make that clear.

File: tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/AlwaysOnProfilerTests.cs [hunk 3]
```diff
@@ -105,8 +105,8 @@ private static bool ContainStackTraceForClassHierarchy(LogRecord logRecord)
             return logRecord.Body.StringValue.Contains(
                 "\tat System.Threading.Thread.Sleep(System.TimeSpan)\n" +
                 "\tat My.Custom.Test.Namespace.ClassD`1.GenericMethodDFromGenericClass(!0, !!0)\n" +
-                "\tat SharedGenericFunction.GenericMethodCFromGenericClass(!0)\n" +
-                "\tat TripleInternalClassB.MethodB(System.String)\n" +
+                "\tat My.Custom.Test.Namespace.GenericClassC`1.GenericMethodCFromGenericClass(!0)\n" +
```
Comment (pjanotti) [inline tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/AlwaysOnProfilerTests.cs:108]: This is definitely looks in line with the symbols that we typically see for .NET 👍

**Review by pjanotti — APPROVED** [review]
LGTM

File: tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp (hunk 1 above)
Comment (Kielek) [reply tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp:320]: Yes, this code is related to internal classes. Comment added.

**Pull request merged**